python/cudf_polars/tests/expressions/test_datetime_basic.py

```diff
@@ -37,24 +40,105 @@ def test_datetime_dataframe_scan(dtype):
     assert_gpu_result_equal(query)
 
 
+datetime_extract_fields = [
+    "year",
+    "month",
+    "day",
+    "weekday",
+    "hour",
+    "minute",
+    "second",
+    "millisecond",
+    "microsecond",
+    "nanosecond",
+]
+
+unsupported_extract_fields = ["millennium"]
+
+
+@pytest.fixture(
+    ids=datetime_extract_fields,
+    params=[methodcaller(f) for f in datetime_extract_fields],
+)
+def field(request):
+    return request.param
+
+
+@pytest.fixture(
+    ids=unsupported_extract_fields,
+    params=[methodcaller(f) for f in unsupported_extract_fields],
+)
+def unsupported_field(request):
+    return request.param
+
+
+def test_datetime_extract(field):
+    ldf = pl.LazyFrame(
+        {
+            "datetimes": pl.datetime_range(
+                datetime.datetime(2020, 1, 1),
+                datetime.datetime(2021, 12, 30),
+                "3mo14h15s11ms33us999ns",
+                eager=True,
+            )
+        }
+    )
+
+    q = ldf.select(field(pl.col("datetimes").dt))
+
+    with pytest.raises(AssertionError):
+        # polars produces int32, libcudf produces int16 for extraction methods
+        # libcudf can lose data here.
+        # https://github.com/rapidsai/cudf/issues/16196
+        assert_gpu_result_equal(q)
+
+    assert_gpu_result_equal(q, check_dtypes=False)
+
+
+def test_datetime_extra_unsupported(unsupported_field):
+    ldf = pl.LazyFrame(
+        {
+            "datetimes": pl.datetime_range(
+                datetime.datetime(2020, 1, 1),
+                datetime.datetime(2021, 12, 30),
+                "3mo14h15s11ms33us999ns",
+                eager=True,
+            )
+        }
+    )
+
+    q = ldf.select(unsupported_field(pl.col("datetimes").dt))
+
+    assert_ir_translation_raises(q, NotImplementedError)
+
+
 @pytest.mark.parametrize(
     "field",
     [
         methodcaller("year"),
-        pytest.param(
-            methodcaller("day"),
-            marks=pytest.mark.xfail(reason="day extraction not implemented"),
-        ),
+        methodcaller("month"),
+        methodcaller("day"),
+        methodcaller("weekday"),
     ],
 )
-def test_datetime_extract(field):
+def test_date_extract(field):
+    ldf = pl.LazyFrame(
+        {
+            "dates": [
+                datetime.date(2024, 1, 1),
+                datetime.date(2024, 10, 11),
+            ]
+        }
+    )
+
     ldf = pl.LazyFrame(
         {"dates": [datetime.date(2024, 1, 1), datetime.date(2024, 10, 11)]}
     )
+
     q = ldf.select(field(pl.col("dates").dt))
 
     with pytest.raises(AssertionError):
-        # polars produces int32, libcudf produces int16 for the year extraction
+        # polars produces int32, libcudf produces int16 for extraction methods
```

Comment:
Same here. Can we upcast the results?

Reply:
Yes, since we're not going to change libcudf behaviour. The right way to do this in `translate.py`. When we observe an extraction function that produces a different dtype in libcudf than polars expects, we should wrap the extraction in a `Cast` node.

That way we can also (in the future) have a configuration flag where (if people are happy not to have matching dtypes), we can avoid doing the casts.

Reply:
I added the cast here as suggested. One drawback is having to maintain a separate list of what functions we don't expect to return the correct type, so that we know what to cast. For instance we need to downcast `month` from `int16` to `int8` to match the polars expected type, because libcudf returns an `int16`. But we _dont_ need to upcast the result of a `.nanosecond()` call because the way we've implemented the processing, we already get an `int32` from the result.

This seems a little weird to me compared to having each function explicitly cast to `self.dtype` during `do_evaluate`, if `self.dtype != result.dtype`. However I don't have a great picture of what putting the cast in the IR _gets us_. I expect @wence might have more to say about the second order effects of putting the cast in the IR rather than the evaluation.

Reply:
By representing the cast in the IR, rather than encoding it implicitly as an evaluation rule, you can potentially (and we do in some cases already) simplify double casts that might exist in the input.

A more general principle is that you want your abstract representation to be as faithful to what will be executed as possible.

Suppose, for example, you want to write a visitor that estimates the memory usage of a query based on some input statistics. As part of that, you'll need to know the dtypes of all the results of the expressions. If those dtypes are only implicit in the evaluation rules, then you need to encode the casting scheme in your new visitor as well, which provides an excellent opportunity for things to get out of date.

Reply:
Thanks for the context!